### DISCUSSION TITLE
Can I use windows-static library to build a DLL in Visual Studio?

### DISCUSSION THREAD

**Discussion by mkaut** [microsoft/vcpkg#22788] [Q&A]
I have an old project that uses QuantLib to build a DLL.
I am testing vcpkg and found that it includes QuantLib, but only in `windows-static` configuration.

To use QuantLib, I set "Use Static Libraries" to "Yes" in the `vcpkg` section of my Visual-Studio project (using VS2019).
However, this gives linker errors of type `error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MTd_StaticDebug' doesn't match value 'MDd_DynamicDebug' in common.obj`, as long as I have "Configuration Type" set to `Dynamic Library (.dll)`.
Switching the configuration type to `Static library (.lib)` fixes the issue - but I need to build a DLL (for use in another project).

Is there some (relatively easy) way to do this?

**Comment by Osyotr** [comment]
By default, static triplets use MT, i.e. static linkage to CRT. You really don't want to mix MT and MD (dynamic linkage to CRT) and that's what your liker is telling you.
To use static libraries with dynamic linkage to CRT, you should set Triplet option to `x64-windows-static-md`.

https://docs.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-170

**Reply by mkaut** [reply]
Thanks - this sounds like the correct approach.
However, even after installing all libraries with the `x64-windows-static-md` triplet, I still haven't managed to make it work: I specified the triplet in the `vcpkg` section of the VS project properties, with `Use static libraries` on either `Yes` or `No`, but get the same linker error nevertheless.

This might be connected to the following error I got during the installation of QuantLib with the triplet:
```
-- Performing post-build validation
Invalid crt linkage. Expected Debug,Dynamic, but the following libs had:

    C:\Users\mkaut\Apps\VCPKG\packages\quantlib_x64-windows-static-md\debug\lib\QuantLib-mt-sgd-x64.lib: Debug,Static

To inspect the lib files, use:
    dumpbin.exe /directives mylibfile.lib
Invalid crt linkage. Expected Release,Dynamic, but the following libs had:

    C:\Users\mkaut\Apps\VCPKG\packages\quantlib_x64-windows-static-md\lib\QuantLib-mt-s-x64.lib: Release,Static

To inspect the lib files, use:
    dumpbin.exe /directives mylibfile.lib
Found 2 post-build check problem(s). To submit these ports to curated catalogs, please first correct the portfile:
    C:\Users\mkaut\Apps\VCPKG\ports\quantlib\portfile.cmake
```

Moreover, I found that I get the same errors also when building an .exe file.
I cannot use more time on this, so I am giving up on using QuantLib via vcpkg...

**Reply by Osyotr** [reply]
@mkaut the problem is in cmake scripts in QuantLib. They hardcode MT for static builds with MSVC.
https://github.com/quantlib/QuantLib/blob/master/cmake/Platform.cmake#L6
